### DISCUSSION TITLE
MessagingMessageConverter is always a StringMessageConverter

### DISCUSSION THREAD

**Discussion by adispennette** [awspring/spring-cloud-aws#1050] [Q&A]
I am trying to us the spring cloud sqs listener for the first time and I have the following config:
```
@Bean
    @Profile("!test")
    fun sqsClient(awsCredentialsProvider: AwsCredentialsProvider): SqsClient {
        return SqsClient.builder().region(Region.of(region))
            .credentialsProvider(awsCredentialsProvider)
            .build()
    }
    @Bean
    @Profile("!test")
    fun sqsAsyncClient(credentialsProvider: AwsCredentialsProvider): SqsAsyncClient {
        return SqsAsyncClient
            .builder()
            .region(Region.of(region))
            .credentialsProvider(credentialsProvider)
            .build()
    }

    @Bean
    fun sqsTemplate(sqsAsyncClient: SqsAsyncClient?, mapper: ObjectMapper): SqsTemplate {
        return SqsTemplate.builder().messageConverter(SqsMessagingMessageConverter()).sqsAsyncClient(sqsAsyncClient).build()
    }
```
The test config for the client is :
```
 @Bean
    fun sqsAsyncClient(awsCredentialsProvider: AwsCredentialsProvider): SqsAsyncClient {
        val sqsEndpoint = AbstractIntegrationTest.localStackContainer.getEndpointOverride(LocalStackContainer.Service.SQS)
        return SqsAsyncClient.builder()
            .endpointOverride(sqsEndpoint)
            .credentialsProvider(awsCredentialsProvider)
            .region(Region.US_WEST_2)
            .build()
    }
```
And the send function using the sqsTemplate is:
```
private fun sendTestMessage(queueUrl: String): SendResult<SendMessageRequest> {
        val json = mapper.writeValueAsString(getS3Event())
        val msgRequest = SendMessageRequest.builder()
            .messageBody(json)
            .queueUrl(queueUrl)
            .build()
        return sqsTemplate.send(QUEUE_NAME, msgRequest)
    }
```
But no matter what I do I can not get the payload converter to use anything other than the string converter. This causes the following error: 
```
java.lang.NullPointerException: payloadMessageConverter returned null message for message GenericMessage
```
Becuase when it get to the type check to see if it supports the type obviously a SendMessageRequest is not a String but Sting is not a valid input for the sqsTemplate.send

Can someone tell me what I am missing that is preventing me from getting the appropriate converter?

**Comment by tomazfernandes** [comment]
Hey @adispennette,

Not sure what you're trying to achieve, and why you're using `SendMessageRequest`.

Just use `sqsTemplate.send(QUEUE_NAME, getS3Event())` and it should work.

Let me know if that works for you.

**Reply by adispennette** [reply]
@tomazfernandes  thanks for the response. I did update the code and initially forgot to update the Generic then realized it and now I get an error message that makes no sense to me and the "docs" don help either
`InvalidAction; see the SQS docs. (Service: Sqs, Status Code: 400, Request ID: 00000000-0000-0000-0000-000000000000)`
I tried looking that up and in the AWS docs found nothing useful but did find another conversation here https://github.com/softwaremill/elasticmq/issues/207 that I am reading now.

**Comment by tomazfernandes** [comment]
Well, this hardly seems like an issue with the framework, so I'll close this issue if you don't mind.

Have you tried looking at the docs for sending messages?
https://docs.awspring.io/spring-cloud-aws/docs/3.1.0/reference/html/index.html#sending-messages

We can reopen the issue if necessary.
